src/controllers/companyController.js

```diff
@@ -0,0 +1,86 @@
+import { companies } from "../models/index.js";
+import NotFoundError from "../errors/notFoundError.js";
+import validateCreateCompany from "../usecases/validateCreateCompany.js";
+
+class CompanyController {
+  static async registerCompany(req, res, next) {
+    try {
+      if (validateCreateCompany(req.body)) {
+        const newCompany = await companies.create(req.body);
+        res.status(201).json({ company: newCompany });
+      }
+    } catch (err) {
+      next(err);
+    }
+  }
+
+  static async listCompanyById(req, res, next) {
+    try {
+      const id = req.params.id;
+      const foundCompany = await companies.findById(id);
+      if (!foundCompany) {
+        next(new NotFoundError("Loja não encontrada!"));
+        return;
+      }
+      res.status(200).json(foundCompany);
+    } catch (err) {
+      next(err);
+    }
+  }
+
+  static async updateCompany(req, res, next) {
+    try {
+      if (validateCreateCompany(req.body, true)) {
+        const id = req.params.id;
+        const updatedCompany = await companies.findByIdAndUpdate(id, req.body, {
+          runValidators: true,
+          new: true,
+        });
+        if (!updatedCompany) {
+          next(new NotFoundError("Loja não encontrada!"));
+          return;
+        }
+        res.status(200).json({ company: updatedCompany });
+      }
```

Comment:
Aqui valem os mesmos comentários que fiz sobre a estrutura de validação do método `registerCompany()`, mas com alguns adendos:

1. A assinatura do método `validateCreateCompany` só recebe um campo, que é os dados que serão validados. Por mais que a aplicação não tenha crashado (o que eu acho uma característica bastante fuleira do JavaScript, mas isso é outro debate hahahaha), é importante que seu método seja chamado conforme a assinatura dele. Se ele só recebe um parâmetro, o ideal é passar só um parâmetro pra ele mesmo.

2. Lembre-se que esse `validateCreateCompany` é utilizado para validar a criação da company. No contrato de criação de uma company, é exigido que os campos sejam obrigatórios e não sejam vazios (name, city, image, document). Porém na atualização não precisamos enviar todos os campos no formulário, apenas os que queremos atualizar (pode ser um, dois ou todos, mas isso fica à critério de quem vai consumir o endpoint).
    
    A partir do momento que a validação de criação de companies é replicada no endpoint de atualização, a API acaba forçando quem for consumir  esse endpoint de informar todos os campos da company, mesmo os que não serão atualizados, caso contrário a atualização não é feita. E além de não ser o comportamento esperado no contrato do endpoint definido na task, você acaba forçando o cliente a enviar dados a mais sem necessidade. Por exemplo, da forma que está, se eu quiser atualizar apenas o nome da company, eu preciso enviar também o documento, a cidade a imagem, mesmo que eu não vá atualizá-los.

   É importante levar em consideração que o usuário pode escolher atualizar só um campo (exemplo: só o `name`, ou o `document`). Sendo assim, é preciso criar um validador só para atualização. Embora sejam as mesmas regras utilizadas na criação de um campo (os campos informados não podem ser vazios e o cnpj da company precisa ser válido), você só vai aplicar essas validações se e somente se o campo for enviado no body de atualização.

Reply:
> A assinatura do método validateCreateCompany só recebe um campo, que é os dados que serão validados. Por mais que a aplicação não tenha crashado (o que eu acho uma característica bastante fuleira do JavaScript, mas isso é outro debate hahahaha), é importante que seu método seja chamado conforme a assinatura dele. Se ele só recebe um parâmetro, o ideal é passar só um parâmetro pra ele mesmo.

My bad Lucão, é resquício de uma outra abordagem que eu queria seguir para lidar com a validação no registro e a na atualização.